provisioning/transport/http/src/http.ts

```diff
@@ -0,0 +1,271 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+'use strict';
+
+import { EventEmitter } from 'events';
+import { RestApiClient, Http as Base } from 'azure-iot-http-base';
+import { errors } from 'azure-iot-common';
+import * as machina from 'machina';
+import { ClientRequest } from 'http';
+
+export type ProvisioningResponseCallback = (err: Error, result?: any) => void;
+
+export class Http extends EventEmitter {
+
+  private _idScope: string;
+  private _registrationId: string;
+  private _body: any;
+  private _forceRegistration: boolean;
+  private _operationId: string;
+
+  private _fsm: machina.Fsm;
+  private _apiVersion: string = '2016-11-14';
+  private _provisioningServer: string = 'https://global.azure-devices-provisioning.net';
+  private _httpTimeout: number = 4000;
+  private _operationStatusPollingInterval: number = 5000;
+  private _pollingTimer: any;
+  private _currentHttpRequest: ClientRequest;
+  private _restApiClient: RestApiClient;
+
+  // TODO: how to get the user agent string into here?
+
+  /**
+   * @private
+   * @constructor
+   * @param config The configuration object.
+   */
+  /* Codes_SRS_NODE_PROVISIONING_HTTP_18_001: [ The `Http` constructor shall accept the following properties:
+  - `idScope` - a string specifiying the scope of the provisioning operations,
+  - `registrationId` - the registration id for the specific device ] */
+  constructor(idScope: string, registrationId: string, httpBase?: Base) {
+    super();
+    this._idScope = idScope;
+    this._registrationId = registrationId;
+
+    this._fsm = new machina.Fsm({
+      namespace: 'http-provisioning',
+      initialState: 'idle',
+      states: {
+        idle: {
+          _onEnter: () => {
+            this._pollingTimer = null;
+            this._currentHttpRequest = null;
+            this._restApiClient = null;
+            this._body = null;
+            this._forceRegistration = false;
+            this._operationId = '';
+          },
+          register: (authorization, forceRegistration, body, callback) => {
+            this._forceRegistration = forceRegistration;
+            this._body = body;
+            this._restApiClient = new RestApiClient({ 'host' : this._provisioningServer , 'sharedAccessSignature' : authorization}, 'azure-node-provisioning', httpBase);
+            this._fsm.transition('registering', callback);
+          },
+          disconnect : (callback) => {
+            // nothing to do.
+            this._safeCallback(callback);
+          }
+        },
+        registering: {
+          _onEnter: (callback) => {
+            this._sendRegistrationRequest((err, responseBody) => {
+              if (err) {
+                this._safeCallback(callback, err);
```

Comment:
>this._safeCallback(callback, err); [](start = 16, length = 34)

since there's a process.nextTick in here we're going to have a weird moment when the callback is "on hold until next tick" while the state machine transitions to idle which is, in case of changes in machina, might lead to some weird side effects. 

Have you considered passing the callback and error in the transition('idle') call and dealing with those in the _onEnter() of the idle state?

Reply:
I was wondering why you did the callback in the _onEnter that way.  I like it.  

---
In reply to: [141481212](https://github.com/Azure/azure-iot-sdk-node/pull/93#discussion_r141481212) [](ancestors = 141481212)